### PR TITLE
#35 - Ranking Backend

### LINKED ISSUE #35: Ranking - Backend
Two versions will be created: one full version for the administration, and a version for students, displaying anonymized data.

### PR DESCRIPTION
**Pull request by PrabuckiDominik**
This should close #35 .

### CODE REVIEW

File: app/Http/Controllers/RankingController.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+{
+    public function index(Quiz $quiz): Response
+    {
+        $this->authorize("viewAdminRanking", $quiz);
+
+        $submissions = QuizSubmission::where("quiz_id", $quiz->id)
+            ->with("user.school")
+            ->get();
+
+        $rankings = $submissions->map(fn($submission) => new RankingResource($submission));
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:25]: Define return type

File: app/Http/Controllers/RankingController.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    public function indexUser(Quiz $quiz): Response
+    {
+        $this->authorize("viewUserRanking", $quiz);
+
+        $submissions = QuizSubmission::where("quiz_id", $quiz->id)
+            ->with("user.school")
+            ->get();
+
+        $rankings = $submissions->map(fn($submission) => new RankingResource($submission));
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:41]: Define return type

File: app/Http/Controllers/RankingController.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+use Inertia\Inertia;
+use Inertia\Response;
+
+class RankingController extends Controller
+{
+    public function index(Quiz $quiz): Response
+    {
+        $this->authorize("viewAdminRanking", $quiz);
+
+        $submissions = QuizSubmission::where("quiz_id", $quiz->id)
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:21]:
```suggestion
        $submissions = QuizSubmission::query()
            ->where("quiz_id", $quiz->id)
```

File: app/Http/Controllers/RankingController.php [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            "quiz" => $quiz,
+            "rankings" => $rankings,
+        ]);
+    }
+
+    public function indexUser(Quiz $quiz): Response
+    {
+        $this->authorize("viewUserRanking", $quiz);
+
+        $submissions = QuizSubmission::where("quiz_id", $quiz->id)
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:37]:
```suggestion
        $submissions = QuizSubmission::query()
            ->where("quiz_id", $quiz->id)
```

File: app/Http/Controllers/RankingController.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    public function publish(Quiz $quiz): RedirectResponse
+    {
+        if (!$quiz->exists) {
+            abort(404);
+        }
+
+        $this->authorize("publish", $quiz);
+
+        $quiz->ranking_published_at = Carbon::now();
+        $quiz->save();
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:58]: Make action for that.

File: app/Http/Controllers/RankingController.php [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    public function unpublish(Quiz $quiz): RedirectResponse
+    {
+        if (!$quiz->exists) {
+            abort(404);
+        }
+
+        $this->authorize("publish", $quiz);
+
+        $quiz->ranking_published_at = null;
+        $quiz->save();
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/RankingController.php:74]: Same here, I would make action.

File: app/Http/Controllers/RankingController.php (hunk 1 above)
Comment (PrabuckiDominik) [reply app/Http/Controllers/RankingController.php]: Done.

File: app/Http/Controllers/RankingController.php (hunk 5 above)
Comment (PrabuckiDominik) [reply app/Http/Controllers/RankingController.php]: Done.

Comment (AleksandraKozubal) [reply app/Http/Controllers/RankingController.php]: Example action: https://github.com/blumilksoftware/toby/blob/main/app/Actions/CreateUserAction.php

**Review by AleksandraKozubal — APPROVED** [review]

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Pull request merged**